perllib/FixMyStreet/App/Controller/Auth/Social.pm

```diff
@@ -142,6 +143,84 @@ sub twitter_callback: Path('/auth/Twitter') : Args(0) {
     $c->forward('oauth_success', [ 'twitter', $info->{id}, $info->{name} ]);
 }
 
+sub oidc : Private {
+    my ($self, $c) = @_;
+
+    my $config = $c->cobrand->feature('oidc_login');
+
+    OIDC::Lite::Client::WebServer::Azure->new(
+        id               => $config->{client_id},
+        secret           => $config->{secret},
+        authorize_uri    => $config->{auth_uri},
+        access_token_uri => $config->{token_uri},
+    );
+}
+
+sub oidc_sign_in : Private {
+    my ( $self, $c ) = @_;
+
+    $c->detach( '/page_error_403_access_denied', [] ) if FixMyStreet->config('SIGNUPS_DISABLED');
+    $c->detach( '/page_error_400_bad_request', [] ) unless $c->cobrand->feature('oidc_login');
+
+    my $oidc = $c->forward('oidc');
+    my $url = $oidc->uri_to_redirect(
+        redirect_uri => $c->uri_for('/auth/OIDC'),
+        scope        => 'openid',
+        state        => 'test',
+        extra        => {
+            response_mode => 'form_post',
+        },
+    );
+
+    my %oauth;
+    $oauth{return_url} = $c->get_param('r');
+    $oauth{detach_to} = $c->stash->{detach_to};
+    $oauth{detach_args} = $c->stash->{detach_args};
+    $c->session->{oauth} = \%oauth;
+    $c->res->redirect($url);
+}
+
+sub oidc_callback: Path('/auth/OIDC') : Args(0) {
```

Comment:
I realise they're all subtly different, but is there any worth in factoring the (now) three out to shared subclasses or something with variants for their differing parts?